flask_resty/jwt.py

```diff
@@ -89,3 +96,53 @@ def get_config_key(self, key):
 
     def get_credentials(self, payload):
         return payload
+
+
+class JwkSetAuthentication(JwtAuthentication):
+    def __init__(self, key_set=None, **kwargs):
+        super(JwkSetAuthentication, self).__init__(**kwargs)
+
+        self.jwks = key_set
+        self.algorithms = get_default_algorithms()
+
+    def get_jwks(self):
+        config = flask.current_app.config
+        return self.jwks \
+            if self.jwks else config[self.get_config_key('key_set')]
+
+    def from_jwk(self, jwk, algorithm):
+        if 'x5c' in jwk:
+            return load_der_x509_certificate(
+                base64.b64decode(jwk['x5c'][0]),
+                default_backend(),
+            ).public_key()
+
+        # awkward
+        return algorithm.from_jwk(json.dumps(jwk))
+
+    def get_key_for_token(self, token):
+        unverified_header = jwt.get_unverified_header(token)
+
+        try:
+            algorithm = self.algorithms[unverified_header['alg']]
```

Comment:
do not use `alg` from here – we can't trust it. instead use `alg` on the JWK. this is just going to have to be a little bit different. we trust `alg` on the JWK but not on the JWT

(though perhaps we should fail if there's a mismatch between the two `alg` fields)

Reply:
Not sure what to do here tho as `alg` is optional on the jwk (pyjwt doesn't even add it). I think in the upstream PR I raise on a mismatch if both are present? That may only be in the encode step tho. Also this what pyjwt does :/

Reply:
it's really unsafe to read it from the header, though. the user controls this value, and it could easily be "no signing" or something else totally screwed up

Reply:
there is a safeguard, the token value is checked against the decode args algorithm, if the user changes it on the token verification will fail, so you cant change it to 'none' unless your server supports that, which would be dumb

Reply:
did you see the auth0 none vuln from a few years ago? what do you think about just requiring an alg in the JWK, or defaulting to a user customizable value (default RS256)?

Reply:
i think on our end we can just require `alg` be in the jwk, but as a generic jwk authentication, that may be too opinionated? IDK, we can certainly require that the config at leasts provides the algo whitelist, that should avoid this vulnerability right? PyJWT sort of requires it, it warns that not providing it is deprecated.